base_layer/service_framework/src/context/future.rs

```diff
@@ -0,0 +1,136 @@
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use super::FinalServiceHandles;
+use crate::context::LazyService;
+use futures::{Future, FutureExt};
+use std::any::Any;
+use tari_shutdown::{Shutdown, ShutdownSignal};
+use tokio::task;
+
+/// Create a Notifier, ServiceInitializerContext pair.
+///
+/// The `Notifier::notify` method will notify all cloned `ServiceHandlesFuture`s
+/// and which will resolve with the collected `ServiceHandles`.
+pub(crate) fn create_context_notifier_pair(shutdown_signal: ShutdownSignal) -> (Shutdown, ServiceInitializerContext) {
+    let trigger = Shutdown::new();
+    let trigger_signal = trigger.to_signal();
+    (trigger, ServiceInitializerContext::new(shutdown_signal, trigger_signal))
+}
+
+/// Contains context for service initialization.
+///
+/// `ServiceInitializerContext` also implements `Future` and resolves to `ServiceHandles` once
+/// it is signaled to do so.
+#[derive(Clone)]
+pub struct ServiceInitializerContext {
+    inner: FinalServiceHandles,
+    ready_signal: ShutdownSignal,
+}
+
+impl ServiceInitializerContext {
+    /// Create a new ServiceInitializerContext.
+    ///
+    /// ## Parameters
+    /// `shutdown_signal` - signal that is provided to services. If this signal is triggered, services should terminate.
+    /// `ready_signal` - indicates that all services are ready. This should be triggered by the `StackBuilder` once all
+    ///                  initializers have run.
+    pub(crate) fn new(shutdown_signal: ShutdownSignal, ready_signal: ShutdownSignal) -> Self {
+        Self {
+            inner: FinalServiceHandles::new(shutdown_signal),
+            ready_signal,
+        }
+    }
+
+    /// Insert a service handle with the given name
+    pub fn register_handle<H>(&self, handle: H)
+    where H: Any + Send + Sync {
+        self.inner.register(handle);
+    }
+
+    /// Retrieve a handle and downcast it to return type and return a copy, otherwise None is returned
+    pub fn get_handle<H>(&self) -> Option<&H>
+    where H: 'static {
+        self.inner.get_handle()
+    }
+
+    /// Get a handle from the given type (`TypeId`) and downcast it to a type `H`.
+    /// If the item does not exist or the downcast fails, a panic occurs
+    pub fn expect_handle<H>(&self) -> H
+    where H: Clone + 'static {
+        self.inner.expect_handle()
+    }
+
+    /// Call the given function with the final handles once this future is ready (`notify_ready` is called).
+    pub fn lazy_service<F, S>(&self, service_fn: F) -> LazyService<F, Self, S>
+    where F: FnOnce(FinalServiceHandles) -> S {
+        LazyService::new(self.clone(), service_fn)
+    }
+
+    /// Spawn a task once handles are ready. The resolved handles are passed into this closure.
+    pub fn spawn_when_ready<F, Fut>(self, f: F) -> task::JoinHandle<Fut::Output>
```

Comment:
Suggestion: maybe `on_started` or `on_ready` is a more clear name?

Reply:
Do you not think it's important to know that you are spawning a new task?